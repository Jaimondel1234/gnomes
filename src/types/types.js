export const types = {
  login: "[Auth] Login",
  logout: "[Auth] Logout",
  loginRequest: "[Auth] Login Request",
};
