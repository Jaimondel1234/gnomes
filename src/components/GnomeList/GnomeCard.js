import React from 'react';
import {
  PanelStyle,
  ImageStyle,
  FlexStyle,
  ContainerStyle,
  ButtonStyle,
  SpanStyle,
} from '../../styles/core';
import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';

/*
  Component that displays the gnome card
*/
export const GnomeCard = ({ ...props }) => {
  const { name, thumbnail } = props;
  const history = useHistory();
  const handleClick = () => {
    history.push('/gnome-detail', { ...props });
  };

  return (
    <PanelStyle
      mr='30px'
      mb='50px'
      maxWidth='900px'
      borderRadius='10px'
      pt='20px'
      mt='10px'
    >
      <FlexStyle
        alignItems='center'
        justifyContent='space-between'
        flexDirection='column'
      >
        <SpanStyle
          color='white'
          fontSize='30px'
          fontWeight='bold'
          p='15px'
          textShadow='2px 2px 4px #000000'
          borderRadius='5px'
          bg='#1E90FF'
        >
          {name}
        </SpanStyle>
        <ContainerStyle p='20px'>
          <ImageStyle
            src={thumbnail}
            width='350px'
            height='300px'
            onError={(e) => {
              e.target.onerror = null;
              e.target.src =
                'http://www.publicdomainpictures.net/pictures/30000/nahled/maple-leaves-background.jpg';
            }}
          />
        </ContainerStyle>
        <ButtonStyle
          primary
          p='12px'
          fontSize='25px'
          fontWeight='bold'
          onClick={handleClick}
          letterSpacing='3px'
          lineHeight='50px'
        >
          ... More
        </ButtonStyle>
      </FlexStyle>
    </PanelStyle>
  );
};

GnomeCard.propTypes = {
  props: PropTypes.shape({
    name: PropTypes.string,
    thumbnail: PropTypes.string,
    age: PropTypes.number,
    genre: PropTypes.string,
    hair_color: PropTypes.string,
    height: PropTypes.number,
    weight: PropTypes.number,
    professions: PropTypes.arrayOf(PropTypes.string),
    friends: PropTypes.arrayOf(PropTypes.string),
  }),
};
