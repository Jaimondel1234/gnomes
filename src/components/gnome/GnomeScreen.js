import React from "react";

export const GnomeScreen = () => {
  return (
    <div>
      <h1>Gnome Screen</h1>
    </div>
  );
};
