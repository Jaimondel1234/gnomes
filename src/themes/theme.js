/* 
  Theme
*/

export const theme = {
  color: {
    main: '#007bff',
  },
};
