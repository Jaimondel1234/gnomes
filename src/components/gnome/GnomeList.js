import React, { useRef, useEffect } from "react";
import { FlexStyle, ScrollContainer } from "../../styles/core";
import { GnomeCard } from "./GnomeCard";
import { useSelector, useDispatch } from "react-redux";
import { loadGnomes } from "../../actions/gnome";

export const GnomeList = () => {
  const dispatch = useDispatch();
  const { gnomes } = useSelector((state) => state.gnome);
  const ref = useRef(null);

  const handleScroll = () => {
    if (ref.current.scrollTop >= 0.75 * ref.current.scrollHeight) {
      dispatch(loadGnomes());
    }
  };
  return (
    <>
      <ScrollContainer ref={ref} onScroll={handleScroll}>
        <FlexStyle
          alignItem="center"
          justifyContent="center"
          flexDirection="row"
          flexWrap="wrap"
        >
          {gnomes &&
            gnomes.map(({ id, ...props }) => <GnomeCard key={id} {...props} />)}
        </FlexStyle>
      </ScrollContainer>
    </>
  );
};
