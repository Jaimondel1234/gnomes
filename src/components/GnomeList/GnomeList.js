import React, { useRef, useState, useEffect } from "react";
import {
  FlexStyle,
  ScrollContainer,
  PlaceHolderStyle,
} from "../../styles/core";
import { GnomeCard } from "./GnomeCard";
import { useSelector, useDispatch } from "react-redux";
import { loadGnomes } from "../../actions/gnome";
import { fill } from "lodash";
import map from "lodash/map";

let emptyArray = fill(new Array(20), 0);

const placeHolderArray = emptyArray.map((value, id) => {
  return `place_holder_${id}`;
});

export const GnomeList = () => {
  const dispatch = useDispatch();
  const { gnomes, loading } = useSelector((state) => state.gnome);
  const ref = useRef(null);

  const handleScroll = () => {
    if (ref.current.scrollTop >= 0.75 * ref.current.scrollHeight) {
      dispatch(loadGnomes());
    }
  };
  return (
    <ScrollContainer ref={ref} onScroll={handleScroll}>
      <FlexStyle
        alignItem="center"
        justifyContent="center"
        flexDirection="row"
        flexWrap="wrap"
      >
        {gnomes &&
          map(gnomes, ({ id, ...props }) => <GnomeCard key={id} {...props} />)}
        {loading &&
          map(placeHolderArray, (value) => (
            <PlaceHolderStyle
              key={value}
              bg="grey"
              mr="30px"
              mb="50px"
              maxWidth="900px"
              borderRadius={10}
              pt="50px"
            />
          ))}
      </FlexStyle>
    </ScrollContainer>
  );
};
