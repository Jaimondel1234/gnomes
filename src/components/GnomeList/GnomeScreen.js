import React, { useEffect } from "react";
import { ContainerStyle, ScrollContainer } from "../../styles/core";
import { GnomeList } from "./GnomeList";
import { loadGnomes } from "../../actions/gnome";
import { useDispatch, useSelector } from "react-redux";

export const GnomeScreen = () => {
  const dispatch = useDispatch();
  const { scrollNumberGnomes } = useSelector((state) => state.gnome);

  useEffect(() => {
    if (scrollNumberGnomes === 0) {
      dispatch(loadGnomes());
    }
  }, [loadGnomes, dispatch, scrollNumberGnomes]);

  return (
    <ContainerStyle mt="5">
      <GnomeList />
    </ContainerStyle>
  );
};
