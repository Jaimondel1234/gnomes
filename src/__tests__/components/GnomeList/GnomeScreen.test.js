import React from "react";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";

import { GnomeScreen } from "../../../components/GnomeList/GnomeScreen";
import { theme } from "../../../themes/theme";
import { types } from "../../../types/types";

describe("Test <GnomeScreen/>", () => {
  const middlewares = [];
  const mockStore = configureStore(middlewares);
  const initState = {
    gnome: { scrollNumberGnomes: 0 },
    auth: {
      uid: "124",
      name: "test",
    },
  };

  const store = mockStore(initState);
  store.dispatch = jest.fn();
  const wrapper = mount(
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <GnomeScreen />
      </ThemeProvider>
    </Provider>
  );
  test("it must be displayed correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("if the number of gnomes is zero and the user is authenticated, new gnomes will be loaded", () => {
    expect(store.dispatch).toHaveBeenCalledWith({
      type: types.loadGnomes,
    });
  });
});
