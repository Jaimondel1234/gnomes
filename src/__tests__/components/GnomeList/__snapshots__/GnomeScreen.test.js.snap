// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <GnomeScreen/> it must be displayed correctly 1`] = `
<div
  className="core__ContainerStyle-qu5lcc-0 kLYPbT"
>
  <div
    className="core__ScrollContainer-qu5lcc-8 dBQIIB"
    onScroll={[Function]}
  >
    <div
      className="core__FlexStyle-qu5lcc-4 eTaOSt"
    />
  </div>
</div>
`;
